diesel/src/sqlite/connection/mod.rs

```diff
@@ -242,6 +242,39 @@ impl SqliteConnection {
         functions::register_aggregate::<_, _, _, _, A>(&self.raw_connection, fn_name)
     }
 
+    /// Register a collation function.
+    ///
+    /// `f` must always return the same answer given the same inputs.
+    ///
+    /// This method will return an error if registering the function fails.
+    ///
+    /// The collation can be specified when creating a table:
+    /// `CREATE TABLE my_table ( str TEXT COLLATE MY_COLLATION )`,
+    /// where `MY_COLLATION` corresponds to `fn_name`.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     let conn = SqliteConnection::establish(":memory:").unwrap();
+    /// conn.register_collation("RUSTNOCASE", |rhs, lhs| {
+    ///     rhs.to_lowercase().cmp(&lhs.to_lowercase())
+    /// })
+    /// # }
+    /// ```
+    pub fn register_collation<F>(&self, fn_name: &str, f: F) -> QueryResult<()>
+    where
+        F: FnMut(&str, &str) -> std::cmp::Ordering + Send + 'static,
```

Comment:
I think a `Fn` instead of `FnMut` communicates the assumption of the callback being deterministic much better.

Reply:
It will not safe-guard against violating that requirement, but that does not impair my use-case so I'll change it.

Reply:
I should word it a bit differently: A callback that is `FnMut` but not `Fn` is quite likely violating the requirement. A callback that is `Fn` needs to actively do bad things (like using random numbers) to violate the requirement.

Reply:
Also with a `Cell` (or friends), but I guess one usually doesn't use that so the point still mostly stands.